### PR TITLE
[SCHEMA] Render schema elements in text

### LINKED ISSUE #591: Standardize structure of entity tables
On a recent call with the NIDM-Terms group, it became clear that the entity tables are somewhat confusingly organized. Namely, auxiliary datatypes flip the conventional pattern without making it clear that the pattern was flipped.

In the current stable version (v1.4.0), there are four patterns:
1. `data type (suffix1 suffix2 etc.)`: Used in most cases.
1. `suffix (data type1/data type 2/etc.)`: Used for what we currently call "auxiliary data types" in the schema (e.g., `channels` files).
1. `data type`: Used in cases where the data type and suffix are the same (e.g., `meg`, `eeg`, `ieeg`). This pattern was removed when we schemafied the entity table, primarily for consistency.
1. `data type (extension1 extension2)`: Only used for DWI with bvec and bval files. I assume this was a typo rather than a design choice, so this pattern was removed when we schemafied the entity table.

In the current latest version (v1.4.1-dev), only the first two patterns remain, but they are still confusing.

I propose that we eliminate the second pattern above (`suffix (data type)`) and reformat it to match the main pattern (i.e., `data type (suffix)`). This will increase the number of rows in the tables (especially the **Encephalography** table), but I don't think that's a major problem now that there are multiple tables.

I also propose that we include the pattern in either the table itself or in the text for the **Entity table** page.

#### Solution 1: Reflect the item in the pattern

We can append `/` to each data type to make it clear it refers to a folder and prepend `_` to each suffix to make it clear it's a suffix.

| Entity               | Subject       | Session       | Task           | Acquisition   | Run           | Recording           |
|----------------------|---------------|---------------|----------------|---------------|---------------|---------------------|
| Format               | `sub-<label>` | `ses-<label>` | `task-<label>` | `acq-<label>` | `run-<index>` | `recording-<label>` |
| beh/<br>(_stim _physio) | REQUIRED      | OPTIONAL      | REQUIRED       | OPTIONAL      | OPTIONAL      | OPTIONAL            |
| beh/<br>(_events _beh)  | REQUIRED      | OPTIONAL      | REQUIRED       | OPTIONAL      | OPTIONAL      |                     |

#### Solution 2: Explicit column header

| Entity               | Subject       | Session       | Task           | Acquisition   | Run           | Recording           |
|----------------------|---------------|---------------|----------------|---------------|---------------|---------------------|
| Modality (Suffixes)               | `sub-<label>` | `ses-<label>` | `task-<label>` | `acq-<label>` | `run-<index>` | `recording-<label>` |
| beh<br>(stim physio) | REQUIRED      | OPTIONAL      | REQUIRED       | OPTIONAL      | OPTIONAL      | OPTIONAL            |
| beh<br>(events beh)  | REQUIRED      | OPTIONAL      | REQUIRED       | OPTIONAL      | OPTIONAL      |                     |

#### Solution 3: Describe the pattern in the text

Currently, we have:

> This section compiles the entities (key-value pairs) described throughout this specification, and establishes a common order within a filename. 
> For example, if a file has an acquisition and reconstruction label, the acquisition entity must precede the reconstruction entity.
> REQUIRED and OPTIONAL entities for a given file type are denoted. 
> Entity formats indicate whether the value is alphanumeric (`<label>`) or numeric (`<index>`).
>
> A general introduction to entities is given in the section on [file name structure](../02-common-principles.md#file-name-structure).

We could have:

> This section compiles the entities (key-value pairs) described throughout this specification, and establishes a common order within a filename. 
> For example, if a file has an acquisition and reconstruction label, the acquisition entity must precede the reconstruction entity.
> REQUIRED and OPTIONAL entities for a given file type are denoted. 
> Entity formats indicate whether the value is alphanumeric (`<label>`) or numeric (`<index>`).
> **Cells in the first column identify the modality and associated suffixes for which each pattern of entities applies.**
>
> A general introduction to entities is given in the section on [file name structure](../02-common-principles.md#file-name-structure).

### PR DESCRIPTION
**Pull request by tsalo**
References #544 and closes #591. Uses the [`mkdocs-macros`](https://mkdocs-macros-plugin.readthedocs.io/en/latest/) plugin to generate certain markdown elements (e.g., the filename patterns) from the schema. This should easily extend to tables and json examples, as well.

Changes proposed:

- Replace manual Entity Table and Entity description file creation scripts with functions that are called by the `mkdocs-macros` plugin within the markdown files.
- Move all auxdatatypes into new groups under the associated data types. This is primarily because I had so much trouble getting a generic filtering function working with the original structure.
- Add standard Python stuff to the gitignore.
- Reorganize Python code into a mini package.
- Add several Python libraries and `macros` as dependencies.
- Replace entities in datatypes files with the actual names instead of their shorthand versions.
- Split the iEEG/EEG/MEG entity table into a Biopotential Amplification (iEEG/EEG) table and an MEG table.

# TODO:

- [x] Reorganize Python code. Even though the `macros` documentation says that the `define_env` function can be in a module, I can't seem to get that working, so it's in its own file at the top level for now.
- [X] Generic filtering function. We need to be able to filter the schema down to specific classes or elements within classes.
    - Something on the order of `pybids` is probably overkill, but I imagine the `pybids` devs would have some insights on this.
    - I added something in 9ce6d7f that is fairly flexible. We should improve it at some point though.
- [X] Updated entity tables. Currently, because of the lack of a filtering function, only the full entity table is generated.
- [X] Combine `.nii` and `.nii.gz` as `.nii[.gz]` in file formats.
- [X] ~~Consider dropping `.json` from the list of extensions in the schema. JSON files are not alternatives to the other extensions. They are required (at some level).~~
    - I've decided to ignore this nuance for now. Will just treat json like any other extension and show it in the format snippets.
- [x] Deal with remark. TravisCI is failing because remark flags the code snippets like `{{ make_filename_template(datatypes=["anat"]) }}` as bad.
- [x] Update contributing information. Fewer steps are needed when updating the schema now, since rendering happens automatically.
- [x] Fix PDF rendering. Currently, it just shows the inline code without the text _produced_ by the code.
- [x] Remove old outputs (currently retained for comparison).

# More To Dos (added by @sappelhoff here for better visibility, but can be done by anybody)

- open issues:
    - [X] eventually turn schemacode into its own package and migrate to separate repo?
        - I've brought this up in #543.
    - [X] add python linter and style check to CI for pdf build code, tools, schemacode
        - I've brought this up in #543.
    - [ ] add specific license for code in this repo ... CC BY 4.0 is more for the spec itself, not so adequate for the code
    - [ ] once this is merged, we'll want to generate "templates" also for files like README, dataset_description.json, and so on (currently not done)
    - [ ] issue to track that files like "scans.tsv/json" and "phenotype/*" get an auto-generated "template" section as well
    - [x] issue for pdf build so that markdown table rows are distinguished with white and gray "fills"
         - Opened #659.

### CODE REVIEW

**Comment by tsalo** [comment]
@yarikoptic check this out! [Here](https://bids-specification--610.org.readthedocs.build/en/610/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#anatomy-imaging-data) is an example filename format example snippet. The first one is autogenerated from the specification, while the second one is the preexisting one. It's far from perfect, but I think it's pretty promising!

**Comment by tsalo** [comment]
I'm marking this ready for review, with a couple of caveats:

1. I still need to update the remark settings so that it will play nicely with `macros`.
1. Some of the filename format snippets for data types with many suffixes look.... not the best. Unfortunately, cleanly combining them without also combining out the `json` files will require either (1) hacky hardcoding or (2) conditional logic implemented in the schema (see #620).
1. What are MEG `_markers` files? They're in the schema but I don't know where they came from.
1. I've kept the original versions of the rendered elements in the specification for now to make comparison easier.

**Comment by sappelhoff** [comment]
> What are MEG _markers files? They're in the schema but I don't know where they came from.

:astonished: did you find a spot in the spec that yields more information? 

I just found this in https://bids-specification.readthedocs.io/en/latest/99-appendices/06-meg-file-formats.html#kityokogawaricoh

Seems like a badly defined suffix ... reminiscent of the `part` -> `split` situation we had a while ago, where an MEG suffix was only defined in an appendix.

**Comment by tsalo** [comment]
It looks like `_markers` has been in the appendix for a couple of years (first mention I could find is #28), but I can't find uses outside of the appendix. I'm not sure where to go from here, except maybe to open a separate issue and tag some MEG folks?

**Comment by sappelhoff** [comment]
> I'm not sure where to go from here, except maybe to open a separate issue and tag some MEG folks?

yes please :+1:

File: .remarkrc
Comment (tsalo) [inline .remarkrc:9]: This doesn't seem to work the way I thought it would. The `remark` test still fails due to this specific warning.

**Comment by yarikoptic** [comment]
@tsalo Sorry for being slow... The example snippet and [rendering](https://bids-specification--610.org.readthedocs.build/en/610/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#anatomy-imaging-data) look absolutely wonderful to me!!! Awesome job!  
The beauty of having it auto-generated is that at some point if we decide to change "rendering" - we would just do that in a single spot -- code.  E.g., as number of entities grows, we might want to us e.g. `sub-<label>[_ses-<label>]_task-<label>[_entity-<value>]*_bold.json` with some "tooltip" when in HTML so when hovering over `[_entity-<value>]*` listing all the entities ... (not here and not now, but in principle, just as an idea).

**Comment by tsalo** [comment]
I think the only remaining concerns are (1) the remark CI and (2) it doesn't look like the PDF is rendering the generated snippets. Does anyone know how to deal with either issue?

**Comment by sappelhoff** [comment]
for the pdf build to succeed we need to add an intermediary build step during the pdf build ci steps ... and in that intermediary step, all the "placeholders" introduced in this pr must be replaced with the actual text.

**Comment by tsalo** [comment]
@sappelhoff I'm not sure where to even start with that. Is there anyone in particular I could/should ask for help with that?

**Comment by yarikoptic** [comment]
there is this python script https://github.com/bids-standard/bids-specification/blob/master/pdf_build_src/process_markdowns.py  where I guess you could just trigger what is needed to be done  probably after `copy_src` call, and just do needed inlining inplace within all those copied .md files

**Comment by sappelhoff** [comment]
> @sappelhoff I'm not sure where to even start with that. Is there anyone in particular I could/should ask for help with that?

@ Arshitha  and @ sebastientourbier habe been helping with the PDF build in the past, if you need a second/third pair of eyes.

And of course @ yarikoptic, but I see that he already replied and what he said is exactly what I meant with my earlier comment.

**Comment by sappelhoff** [comment]
@tsalo if I understand it correctly, the way to go from the snippets like  `{{ make_filename_template(datatypes=["anat"]) }}` to the actual text is handled via a Mkdocs Plugin during the Mkdocs build, right?

I think what we need to make the PDF build work, is a python script to "manually" change the snippets to the actually text through calling that python script.

We'd then integrate that script in the PDF build flow.

File: src/04-modality-specific-files/03-electroencephalography.md
Comment (sappelhoff) [inline src/04-modality-specific-files/03-electroencephalography.md:282]: this produces a template that includes a `electrodes.json`, however there is no such file. Metadata for `electrodes.tsv` is specified in `coordsystem.json`:

image

Comment (sappelhoff) [reply src/04-modality-specific-files/03-electroencephalography.md:282]: also, I think that the `ses-label` would need to go on a new line (and indentation appropriately fixed)

Lastly, we should have a discussion on whether we do want the square brackets to indicate optional Req. Level. or not --> and then apply that consistently. It's currently not documented anywhere. (This should be a new issue)

Comment (yarikoptic) [reply src/04-modality-specific-files/03-electroencephalography.md:282]: re `[]` -- aren't they are used already throughout bids specification (see e.g. [anatomical](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#anatomy-imaging-data))? if anything we indeed just need clarification somewhere in "common principles", but IMHO it isn't specific to this PR anyhow

Comment (yarikoptic) [reply src/04-modality-specific-files/03-electroencephalography.md:282]: re `_electrodes.json` ... eh indeed, there is no `_electrodes.json` but `_electrodes.tsv` must be accompanied with `_coordsystem.json` [apparently](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/03-electroencephalography.html#electrodes-description-_electrodestsv).  Not sure if that is anyhow expressible in the schema ATM.
Also I noted in there a use of `[]` to annotate that an entire file is "optional" (`[sub-<label>[_ses-<label>][_acq-<label>]_coordsystem.json]`) -- not sure if we should do that.  Majority of the files are optional or with some tricky dependencies like here `_electrodes.tsv` requires `_coordsystem.json`

Comment (sappelhoff) [reply src/04-modality-specific-files/03-electroencephalography.md:282]: > eh indeed, there is no _electrodes.json but _electrodes.tsv must be accompanied with _coordsystem.json apparently.

yes. 

> Also I noted in there a use of [] to annotate that an entire file is "optional"

yup, that's what I meant. I think we should stop doing that, and document [] for option ENTITIES in a filename.

(in a separate PR)

Comment (tsalo) [reply src/04-modality-specific-files/03-electroencephalography.md:282]: > also, I think that the ses-label would need to go on a new line (and indentation appropriately fixed)

This is done inconsistently in the specification. In MEG, iEEG, and EEG, the templates have a mix of two-space and four-space tabs and `ses` is on a separate line from `sub`, but in the rest of the sections, they use four-space tabs and `ses` is on the same line as `sub`. I went with the latter, but can change if you prefer the former.

>eh indeed, there is no _electrodes.json but _electrodes.tsv must be accompanied with _coordsystem.json apparently. Not sure if that is anyhow expressible in the schema ATM.

I have an open issue about that kind of logic (#620), but no actual solution just yet. The same goes for labeling entire files as optional. In the mean time, I can just remove `json` as a valid extension for `electrodes` files.

Comment (tsalo) [reply src/04-modality-specific-files/03-electroencephalography.md:282]: Actually I think fixing `electrodes.json` should probably go in a separate PR since it's an issue with the actual schema. I'll open one ASAP.

**Comment by sappelhoff** [comment]
continuation of thread starting here: https://github.com/bids-standard/bids-specification/pull/610#discussion_r514150118

> but can change if you prefer the former

most important to me would be consistency.
For everything else I only have my personal opinion to prefer this style:

````
Template:

```Text
sub-<label>
    [ses-<label>/]
        fmap/
            sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_phase1.nii[.gz]
            sub-<label>[_ses-<label>][_acq-<label>][_run-<index>]_phase1.json
```
````

That is:
- a heading that says `Template:`
- then a "text" codeblock in which the rest is nested
- every directory level on a new line and indented
- indentation: 4 spaces
- optional parts of a filename indicated by square brackets
- optional parts of the directory structure indicated by square brackets

note that I would stop indicating optional files by square brackets. I.e., not `[sub-<label>_coordystem.json]` to indicate that coordsystem.json is optional. For the reason that we have too many files that are conditionally optional, and that the [] syntax does not allow to indicate other requirement levels (optional, recommended, required, conditionally required). As also outlined by @yarikoptic before.

And perhaps this PR would be good to implement these changes after all?

More opinions perhaps by @effigies?

File: src/04-modality-specific-files/03-electroencephalography.md
Comment (sappelhoff) [reply src/04-modality-specific-files/03-electroencephalography.md:282]: let's continue discussion in the main thread: https://github.com/bids-standard/bids-specification/pull/610#issuecomment-718852640 ... this code-review-comment thread can easily become "invisible"

File: .remarkrc
Comment (sappelhoff) [reply .remarkrc:9]: did you try adding:

```suggestion
    ["lint-no-shortcut-reference-link", false]
```
?

**Review by sappelhoff — COMMENTED** [review]
@tsalo could you also have a shot at the templates in https://github.com/bids-standard/bids-specification/blob/master/src/03-modality-agnostic-files.md ?

File: .remarkrc
Comment (tsalo) [reply .remarkrc]: That worked! Thanks!

**Comment by tsalo** [comment]
> @tsalo could you also have a shot at the templates in https://github.com/bids-standard/bids-specification/blob/master/src/03-modality-agnostic-files.md ?

My filtering and rendering functions are, unfortunately, not advanced enough to grab the top-level files... I spent some time today trying to figure out how to improve them, but couldn't. I'm hoping that folks will be able to help improve them iteratively over time.

**Comment by tsalo** [comment]
@sappelhoff @yarikoptic I wrote a function to evaluate the `macros` calls and replace the text in the PDF versions and... it worked! :tada::tada::tada:

**Comment by sappelhoff** [comment]
> I wrote a function to evaluate the macros calls and replace the text in the PDF versions and... it worked! 

Awesome! I'll make a thorough review later today!

**Comment by sappelhoff** [comment]
PS: #653 finally fixed https://github.com/bids-standard/bids-specification/pull/610#issuecomment-702927730, which came up in this thread - so I don't know whether that resolved something for you or whether it adds a new to-do for you @tsalo :slightly_smiling_face:

**Review by sappelhoff — COMMENTED** [review]
This looks like a great PR to me. I did a first pass over the files and left many comments, mostly about the docstring style.

Sorry if this seems nitpicky, but I think it's important to enforce consistency and exhaustive documentation here so that we make our future lives easier.

# to do
- [ ] I did not yet compare the "outputs" of these changes in the new rendered spec to the previously rendered spec
- [ ] need to open issue so that this rendering gets also done for the top level files like README, dataset_description.json, and so on

# questions

- is all this code covered by the CC BY 4.0 license we have at the root of the repo? Or should we add a specific license for the code?

- would it perhaps make sense to install a python linter and style enforcer into our CIs, so that the python *code* is checked?

**Comment by sappelhoff** [comment]
I reviewed:

- EEG
- MEG
- iEEG
- Behavioral

to do

- [ ] mri
- [ ] entity table
- [ ] entity list


# iEEG electrodes.tsv should accept acq entity

here the "old" rendering is correct, the new one is not. see #650 

image

# EEG electrodes.tsv should be the same as iEEG

i.e., accept "space" (for consistency), but do not accept "run"

image


# "markers.json" in MEG need to be fixed

it's either `*_markers.sqd` or `*_markers.mrk`, see #653 note that `_markers` is only supported for "KIT/Yokogawa/Ricoh" MEG data

image

# MEG headshape.ext files do not have a specific extension

i.e., all possible kinds of extensions are allowed, I am afraid (not only .pos) ... look at this issue: https://github.com/bids-standard/bids-specification/issues/20

spoiler alert: It's really an old but still open wound

image

**Comment by sappelhoff** [comment]
more questions

--> what about physio, genetics, and task events? Did you want to add the automatic "template" section for those at a later PR?

--> apart from the top level files (participants, README, LICENSE, CHANGES, ...) we also have "scans files" and "phenotype" files listed under modality agnostic. For top level files you said that rendering them is currently not (yet) possible ... is it the same situation for scans and phenotype? --> or is that a separate issue that requires a separate issue thread (and follow up PR) to be opened?

**Comment by tsalo** [comment]
> --> what about physio, genetics, and task events? Did you want to add the automatic "template" section for those at a later PR?

There is currently no definition of a "match" in the schema. I do bring it up as something we want to support in https://github.com/bids-standard/bids-specification/issues/620#issuecomment-702766358 though.

> --> apart from the top level files (participants, README, LICENSE, CHANGES, ...) we also have "scans files" and "phenotype" files listed under modality agnostic. For top level files you said that rendering them is currently not (yet) possible ... is it the same situation for scans and phenotype? --> or is that a separate issue that requires a separate issue thread (and follow up PR) to be opened?

I think we need a separate PR to support other filetypes.

>"markers.json" in MEG need to be fixed
Done!

>EEG electrodes.tsv should be the same as iEEG
Done

>iEEG electrodes.tsv should accept acq entity
Done

>MEG headshape.ext files do not have a specific extension
I had to add a special "*" character for an unrestricted extension, but I think it works.

**Review by sappelhoff — COMMENTED** [review]
the entity table was *really* hard to review and also a bit confusing. I think our goal with the review here should be that the new auto-generated entity table looks equal to the previous one --> if we find issues, we should raise them as separate GH issues.

and overall, I think the auto-rendered entity table needs a couple of more eyes to be reviewed - I can imagine that I missed something.


# entity table

mri: did DWI disappear from the auto-rendered entity table?


# entity list

for the split entity, the tool "MNE" is mentioned. This should be changed to MNE-Python, and a link to "https://mne.tools" should be added.

https://github.com/bids-standard/bids-specification/blob/master/src/schema/entities.yaml

# pdf

looks okay --> the mri entity table is a bit too large to be rendered nicely, but that's okay IMHO

what would be a great improvement would be to have every second row of the table get a light gray fill ... to better distinguish rows. Not sure if that's possible, and it's certainly out of the scope of this PR.

# mri templates

looks all fine to me

what's a pity is that the internal links from entity to the entity list do not work, we track that in #596 - sounds like a fairly easy to solve issue for 80% of the links. Let's see whether someone finds time soonish

**Comment by sappelhoff** [comment]
overall my review is done now - if you address my remarks @tsalo, then we'd need one or two more reviews by other people and can merge it, I think.

**Comment by tsalo** [comment]
> did DWI disappear from the auto-rendered entity table?

🤦 yes, yes it did. Thank you for catching that!

> for the split entity, the tool "MNE" is mentioned. This should be changed to MNE-Python, and a link to "https://mne.tools" should be added.

Done!

> the mri entity table is a bit too large to be rendered nicely, but that's okay IMHO

I think that is a pre-existing issue, unfortunately.

> what would be a great improvement would be to have every second row of the table get a light gray fill ... to better distinguish rows. Not sure if that's possible, and it's certainly out of the scope of this PR.

The new code just returns a MarkDown table, so I think that's something new that we'll need to add to the PDF build step. Probably worth a new issue.

I'll try to open the relevant new issues today, so that the ideas that have popped up from this PR don't get lost. Thank you for reviewing and helping out so much.

**Comment by sappelhoff** [comment]
Just a reminder to us that we should merge this after ASL and qMRI, but before PET ... so that we can finally start reaping the benefits.

File: src/schema/datatypes/eeg.yaml
Comment (sappelhoff) [inline src/schema/datatypes/eeg.yaml:41]: Actually, JSON is permitted, I think I was wrong about this in the past

```suggestion
     - .json
```

**Review by sappelhoff — APPROVED** [review]
apart from one comment, LGTM

**Comment by sappelhoff** [comment]
@effigies @emdupre @Remi-Gau @yarikoptic we would like to merge this PR but would feel more comfortable with at least one more review.

**Review by yarikoptic — APPROVED** [review]
I have skimmed through it, saw some of my requested (forgot the medium we communicated through with @tsalo awhile back) changes reflected .  This review includes some of my comments (e.g. eventually should be RFed so there is no `sys.path`) which I never submitted (will not remove them) but they can be ignored.
This is a HUGE work and progress, thank you @tsalo for leading it, and everyone who contributed.  Let's get it merged and work on top of that ;)

**Comment by yarikoptic** [comment]
Note that before merge the "Remove old outputs (currently retained for comparison)." should still be done (in entities table).

File: tools/mkdocs_macros_bidsschema/main.py
Comment (tsalo) [inline tools/mkdocs_macros_bidsschema/main.py:10]: Sorry @yarikoptic, but this is the current version of what you commented on in https://github.com/bids-standard/bids-specification/pull/610#discussion_r528005803. Should I add `__init__.py` to this line?

File: CONTRIBUTING.md
Comment (tsalo) [reply CONTRIBUTING.md:505]: will do

File: CONTRIBUTING.md
Comment (tsalo) [reply CONTRIBUTING.md:511]: You're right, the CI will render the files, but the CI _won't_ fail if the code fails.

Comment (yarikoptic) [reply CONTRIBUTING.md:512]: oh, how come is that?  because it is too deep (invoked through remake etc) and something just swallows its failures?  Ideally we should fail if anything fails in the code... if not possible "legitimately", could be done by wrapping `__main__` invocation in try/except and then creating some sentinel file and then checking after invocation of the outside tool that there is no such file.

just a comment/idea: should not stop this PR from being merged!

Comment (tsalo) [reply CONTRIBUTING.md:512]: I think it's a problem with the `mkdocs-macros` plugin.

File: pdf_build_src/process_markdowns.py
Comment (tsalo) [reply pdf_build_src/process_markdowns.py]: It looks like adding `__init__.py` broke things.

Comment (sappelhoff) [reply pdf_build_src/process_markdowns.py]: @tsalo I suggest to revert for now. We can smooth and streamline later on.

Comment (tsalo) [reply pdf_build_src/process_markdowns.py]: Will do.

**Review by sappelhoff — APPROVED** [review]
as discussed in maintainers meeting: let's remove the "old template texts" and then merge

**Pull request merged**
